soci/gzip_indexer.go

```diff
@@ -0,0 +1,147 @@
+	return SpanID(C.pt_index_from_ucmp_offset(i.index, C.long(offset)))
+}
+
+// HasBits wraps `C.has_bits` and returns true if any data is contained in the previous span.
+func (i *GzipIndexer) HasBits(spanID SpanID) bool {
+	return C.has_bits(i.index, C.int(spanID)) != 0
+}
+
+// GetCompressedOffset wraps `C.get_comp_off` and returns the offset for the span in the compressed stream.
+func (i *GzipIndexer) GetCompressedOffset(spanID SpanID) FileSize {
+	return FileSize(C.get_comp_off(i.index, C.int(spanID)))
+}
+
+// GetUncompressedOffset wraps `C.get_uncomp_off` and returns the offset for the span in the uncompressed stream.
+func (i *GzipIndexer) GetUncompressedOffset(spanID SpanID) FileSize {
+	return FileSize(C.get_ucomp_off(i.index, C.int(spanID)))
+}
+
+// ExtractDataFromBuffer wraps the call to `C.extract_data_from_buffer`, which takes in the compressed bytes
+// and returns the decompressed bytes.
+func (i *GzipIndexer) ExtractDataFromBuffer(compressedBuf []byte, uncompressedSize, uncompressedOffset FileSize, spanID SpanID) ([]byte, error) {
+	bytes := make([]byte, uncompressedSize)
+	ret := C.extract_data_from_buffer(
+		unsafe.Pointer(&compressedBuf[0]),
+		C.off_t(len(compressedBuf)),
+		i.index,
+		C.off_t(uncompressedOffset),
+		unsafe.Pointer(&bytes[0]),
+		C.off_t(uncompressedSize),
+		C.int(spanID),
+	)
+	if ret <= 0 {
+		return bytes, fmt.Errorf("error extracting data; return code: %v", ret)
+	}
+
+	return bytes, nil
+}
+
+// ExtractData wraps `C.extract_data` and returns the decompressed bytes given the name of the .tar.gz file,
+// offset and the size in uncompressed stream.
+func (i *GzipIndexer) ExtractData(fileName string, uncompressedSize, uncompressedOffset FileSize) ([]byte, error) {
+	cstr := C.CString(fileName)
+	defer C.free(unsafe.Pointer(cstr))
+	bytes := make([]byte, uncompressedSize)
+	ret := C.extract_data(cstr, i.index, C.off_t(uncompressedOffset), unsafe.Pointer(&bytes[0]), C.int(uncompressedSize))
+	if ret <= 0 {
+		return nil, fmt.Errorf("unable to extract data; return code = %v", ret)
+	}
+
+	return bytes, nil
+}
+
+// GetSpanIndicesForFile wraps `C.span_indices_for_file` and returns IDs of starting and ending given start and end offsets.
+func (i *GzipIndexer) GetSpanIndicesForFile(startOffset, endOffset FileSize) (SpanID, SpanID, error) {
+	var indexStart SpanID
+	var indexEnd SpanID
+	ret := C.span_indices_for_file(i.index, C.off_t(startOffset), C.off_t(endOffset), unsafe.Pointer(&indexStart), unsafe.Pointer(&indexEnd))
+	if int(ret) <= 0 {
+		return 0, 0, fmt.Errorf("cannot get the span indices for file with start and end offset: %d, %d; gzip error: %v", startOffset, endOffset, ret)
+	}
+	return indexStart, indexEnd, nil
+}
+
+func NewGzipIndexer(indexData []byte) (*GzipIndexer, error) {
+	index := C.blob_to_index(unsafe.Pointer(&indexData[0]))
+	if index == nil {
+		return nil, fmt.Errorf("cannot convert blob to gzip_index")
+	}
```

Comment:
Would also recommend adding a finalizer here that will call `Close` when the object here is GC-ed.

Reply:
Not the best idea. Tests show that this causes panics due to access memory violation. That is why the API was designed in the way to let the caller do `defer indexer.Close()`. Moreover, it's already the pattern with IO operations.

Reply:
Yes, but most IO operations are scoped to a function, so it makes sense to `defer`. The Indexer might not.